### 1. Unión automatizada de los diferentes ficheros de entrada.

In [67]:
import pandas as pd
import numpy as np
pd.options.display.max_columns = None

In [68]:
archivo_txt = pd.read_csv("data/archivo_txt.csv", index_col=0)
archivo_txt.head()

index         Q3                 Q4                       Q5          Q6  \
0      0      India  Bachelor’s degree                    Other  5-10 years   
1      1  Indonesia    Master’s degree  Program/Project Manager   20+ years   
2      2   Pakistan    Master’s degree        Software Engineer   1-3 years   
3      3     Mexico    Doctoral degree       Research Scientist   20+ years   
4      4      India    Doctoral degree                    Other   < 1 years   

       Q8                                                Q11  \
0  Python                                           A laptop   
1  Python  A cloud computing platform (AWS, Azure, GCP, h...   
2  Python                                           A laptop   
3  Python  A cloud computing platform (AWS, Azure, GCP, h...   
4  Python  A cloud computing platform (AWS, Azure, GCP, h...   

                  Q13                                    Q15  \
0           2-5 times                             5-10 years   
1               Never                               < 1 year   
2               Never  I do not use machine learning methods   
3  More than 25 times                             5-10 years   
4               Never                            10-20 years   

                         Q20                   Q21  Q22  \
0  Manufacturing/Fabrication      50-249 employees  3-4   
1  Manufacturing/Fabrication  1000-9,999 employees  1-2   
2        Academics/Education  1000-9,999 employees    0   
3        Academics/Education  1000-9,999 employees    0   
4        Academics/Education      50-249 employees  5-9   

                                                 Q23            Q25  \
0                      No (we do not use ML methods)  25,000-29,999   
1  We are exploring ML methods (and may one day p...  60,000-69,999   
2                                      I do not know         $0-999   
3                                      I do not know  30,000-39,999   
4  We use ML methods for generating insights (but...  30,000-39,999   

            Q26     Q33  Q35  \
0     $100-$999     NaN  NaN   
1     $0 ($USD)     NaN  NaN   
2     $0 ($USD)  MySQL   NaN   
3     $0 ($USD)     NaN  NaN   
4  $1000-$9,999     NaN  NaN   

                                                 Q41                  Q7  \
0  Local development environments (RStudio, Jupyt...           Python, R   
1    Advanced statistical software (SPSS, SAS, etc.)   SQL, C, C++, Java   
2  Basic statistical software (Microsoft Excel, G...   Python, C++, Java   
3  Local development environments (RStudio, Jupyt...              Python   
4  Local development environments (RStudio, Jupyt...   Python, C, MATLAB   

                                 Q9                 Q12  \
0                       Vim / Emacs  Google Cloud TPUs    
1        Notepad++,Jupyter Notebook                None   
2  PyCharm ,Jupyter Notebook, Other               Other   
3           Spyder,Jupyter Notebook        NVIDIA GPUs    
4   Spyder,MATLAB ,Jupyter Notebook                None   

                                                 Q14  \
0   Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ...   
1                                        Matplotlib    
2                                        Matplotlib    
3                                        Matplotlib    
4             Matplotlib ,Seaborn ,Ggplot / ggplot2    

                                 Q16  \
0  Scikit-learn , TensorFlow ,Caret    
1                      Scikit-learn    
2                                NaN   
3  Scikit-learn , TensorFlow ,Keras    
4   Scikit-learn ,PyTorch ,LightGBM    

                                                 Q17  \
0  Linear or Logistic Regression, Decision Trees ...   
1  Linear or Logistic Regression, Decision Trees ...   
2                                                NaN   
3   Dense Neural Networks (MLPs, etc), Convolutio...   
4  Linear or Logistic Regression, Decision Trees ...   

                                                 Q24    

In [25]:
archivo_xml = pd.read_csv("data/archivo_xml.csv", index_col=0)
archivo_xml.head()

level_0  index  time    age gender
0        0      0   910  50-54    Man
1        1      1   784  50-54    Man
2        2      2   924  22-24    Man
3        3      3   575  45-49    Man
4        4      4   781  45-49    Man

In [69]:
data_remaning_kaggle = pd.read_csv("data/data_remaning_kaggle.csv", index_col=0)
data_remaning_kaggle.head()

index                                         Q18_Part_1  \
0      0  General purpose image/video tools (PIL, cv2, s...   
1      1                                                NaN   
2      2                                                NaN   
3      3  General purpose image/video tools (PIL, cv2, s...   
4      4                                                NaN   

                                          Q18_Part_2  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  Image segmentation methods (U-Net, Mask R-CNN,...   
4                                                NaN   

                                          Q18_Part_3  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  Object detection methods (YOLOv3, RetinaNet, etc)   
4                                                NaN   

                                          Q18_Part_4 Q18_Part_5 Q18_Part_6  \
0                                                NaN        NaN        NaN   
1                                                NaN        NaN        NaN   
2                                                NaN        NaN        NaN   
3  Image classification and other general purpose...        NaN        NaN   
4                                                NaN        NaN        NaN   

  Q18_OTHER                                         Q19_Part_1 Q19_Part_2  \
0       NaN  Word embeddings/vectors (GLoVe, fastText, word...        NaN   
1       NaN                                                NaN        NaN   
2       NaN                                                NaN        NaN   
3       NaN                                                NaN        NaN   
4       NaN                                                NaN        NaN   

  Q19_Part_3 Q19_Part_4 Q19_Part_5 Q19_OTHER Q27_A_Part_1       Q27_A_Part_2  \
0        NaN        NaN        NaN       NaN          NaN                NaN   
1        NaN        NaN        NaN       NaN          NaN                NaN   
2        NaN        NaN        NaN       NaN          NaN                NaN   
3        NaN        NaN       None       NaN          NaN                NaN   
4        NaN        NaN        NaN       NaN          NaN   Microsoft Azure    

                    Q27_A_Part_3 Q27_A_Part_4 Q27_A_Part_5 Q27_A_Part_6  \
0   Google Cloud Platform (GCP)           NaN          NaN          NaN   
1                            NaN          NaN          NaN          NaN   
2                            NaN          NaN          NaN          NaN   
3                            NaN          NaN          NaN          NaN   
4                            NaN          NaN          NaN          NaN   

  Q27_A_Part_7 Q27_A_Part_8 Q27_A_Part_9 Q27_A_Part_10 Q27_A_Part_11  \
0          NaN          NaN          NaN           NaN           NaN   
1          NaN          NaN          NaN           NaN           NaN   
2          NaN          NaN          NaN           NaN          None   
3          NaN          NaN          NaN           NaN           NaN   
4          NaN          NaN          NaN           NaN           NaN   

  Q27_A_OTHER  Q28 Q29_A_Part_1                        Q29_A_Part_2  \
0         NaN  NaN          NaN                                 NaN   
1         NaN  NaN          NaN                                 NaN   
2         NaN  NaN          NaN                                 NaN   
3         NaN  NaN          NaN                                 NaN   
4         NaN  NaN          NaN   Microsoft Azure Virtual Machines    

                    Q29_A_Part_3 Q29_A_Part_4 Q29_A_OTHER  \
0   Google Cloud Compute Engine           NaN         NaN   
1                            NaN          NaN         NaN   
2                            NaN          N

In [70]:
exported_db_data = pd.read_csv("data/exported_db_data.csv")
exported_db_data.head()

index         Q10_Part_1       Q10_Part_2 Q10_Part_3 Q10_Part_4 Q10_Part_5  \
0      0                NaN  Colab Notebooks        NaN        NaN        NaN   
1      1   Kaggle Notebooks  Colab Notebooks        NaN        NaN        NaN   
2      2   Kaggle Notebooks              NaN        NaN        NaN        NaN   
3      3                NaN  Colab Notebooks        NaN        NaN        NaN   
4      4                NaN              NaN        NaN        NaN        NaN   

  Q10_Part_6 Q10_Part_7 Q10_Part_8 Q10_Part_9 Q10_Part_10  \
0        NaN        NaN        NaN        NaN         NaN   
1        NaN        NaN        NaN        NaN         NaN   
2        NaN        NaN        NaN        NaN         NaN   
3        NaN        NaN        NaN        NaN         NaN   
4        NaN        NaN        NaN        NaN         NaN   

            Q10_Part_11 Q10_Part_12 Q10_Part_13 Q10_Part_14 Q10_Part_15  \
0                   NaN         NaN         NaN         NaN         NaN   
1                   NaN         NaN         NaN         NaN         NaN   
2                   NaN         NaN         NaN         NaN         NaN   
3                   NaN         NaN         NaN         NaN         NaN   
4  Google Cloud Datalab         NaN         NaN         NaN         NaN   

  Q10_Part_16 Q10_OTHER  
0         NaN       NaN  
1         NaN       NaN  
2         NaN       NaN  
3         NaN       NaN  
4         NaN       NaN

Vamos a ver un pequeño resumen de los tipos de datos de cada columna, nombre, valores faltantes de cada uno de los DataFrame para tener más información.

In [29]:
archivo_txt.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25973 entries, 0 to 25972
Data columns (total 27 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   index   25973 non-null  int64 
 1   Q3      25973 non-null  object
 2   Q4      25973 non-null  object
 3   Q5      25973 non-null  object
 4   Q6      25973 non-null  object
 5   Q8      24940 non-null  object
 6   Q11     24720 non-null  object
 7   Q13     24403 non-null  object
 8   Q15     23754 non-null  object
 9   Q20     16325 non-null  object
 10  Q21     16251 non-null  object
 11  Q22     16126 non-null  object
 12  Q23     15952 non-null  object
 13  Q25     15391 non-null  object
 14  Q26     15183 non-null  object
 15  Q33     4656 non-null   object
 16  Q35     2237 non-null   object
 17  Q41     18489 non-null  object
 18  Q7      24941 non-null  object
 19  Q9      24934 non-null  object
 20  Q12     24582 non-null  object
 21  Q14     23970 non-null  object
 22  Q16     19624 non-null

In [30]:
archivo_xml.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25973 entries, 0 to 25972
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   level_0  25973 non-null  int64 
 1   index    25973 non-null  int64 
 2   time     25973 non-null  int64 
 3   age      25973 non-null  object
 4   gender   25973 non-null  object
dtypes: int64(3), object(2)
memory usage: 1.2+ MB


In [31]:
data_remaning_kaggle.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25973 entries, 0 to 25972
Columns: 213 entries, index to Q38_B_OTHER
dtypes: float64(8), int64(1), object(204)
memory usage: 42.4+ MB


In [32]:
exported_db_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25973 entries, 0 to 25972
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   index        25973 non-null  int64 
 1   Q10_Part_1   9507 non-null   object
 2   Q10_Part_2   9792 non-null   object
 3   Q10_Part_3   1062 non-null   object
 4   Q10_Part_4   209 non-null    object
 5   Q10_Part_5   1770 non-null   object
 6   Q10_Part_6   175 non-null    object
 7   Q10_Part_7   1067 non-null   object
 8   Q10_Part_8   753 non-null    object
 9   Q10_Part_9   348 non-null    object
 10  Q10_Part_10  2051 non-null   object
 11  Q10_Part_11  1794 non-null   object
 12  Q10_Part_12  590 non-null    object
 13  Q10_Part_13  251 non-null    object
 14  Q10_Part_14  260 non-null    object
 15  Q10_Part_15  147 non-null    object
 16  Q10_Part_16  7174 non-null   object
 17  Q10_OTHER    737 non-null    object
dtypes: int64(1), object(17)
memory usage: 3.6+ MB


---

Después de su lectura comprobamos que todos tienen el mismo número de registros: 25973..

Por lo deducimos que todos los DataFrames hay que unirlos por la columna index ya que vemos que todos tienen la misma numeración.

Ahora vamos a hacer la unión de los cuatro DataFrames en uno solo y así empezar a contestar a nuestras preguntas. Para eso vamos a almacenar en una variable (array) el nombre de las columnas y aplicar el método intersect1d de Numpy para obtener los valores de las columnas que tienen en común en la combinación de tablas dos a dos.

In [33]:
columnas_data_txt_clean = archivo_txt.columns
columnas_data_txt_clean

Index(['index', 'Q3', 'Q4', 'Q5', 'Q6', 'Q8', 'Q11', 'Q13', 'Q15', 'Q20',
       'Q21', 'Q22', 'Q23', 'Q25', 'Q26', 'Q33', 'Q35', 'Q41', 'Q7', 'Q9',
       'Q12', 'Q14', 'Q16', 'Q17', 'Q24', 'Q32', 'Q34'],
      dtype='object')

In [34]:
columnas_archivo_xml_clean = archivo_xml.columns
columnas_archivo_xml_clean

Index(['level_0', 'index', 'time', 'age', 'gender'], dtype='object')

In [35]:
columnas_data_remaning_kaggle = data_remaning_kaggle.columns
columnas_data_remaning_kaggle

Index(['index', 'Q18_Part_1', 'Q18_Part_2', 'Q18_Part_3', 'Q18_Part_4',
       'Q18_Part_5', 'Q18_Part_6', 'Q18_OTHER', 'Q19_Part_1', 'Q19_Part_2',
       ...
       'Q38_B_Part_3', 'Q38_B_Part_4', 'Q38_B_Part_5', 'Q38_B_Part_6',
       'Q38_B_Part_7', 'Q38_B_Part_8', 'Q38_B_Part_9', 'Q38_B_Part_10',
       'Q38_B_Part_11', 'Q38_B_OTHER'],
      dtype='object', length=213)

In [36]:
columnas_exported_db_data = exported_db_data.columns
columnas_exported_db_data

Index(['index', 'Q10_Part_1', 'Q10_Part_2', 'Q10_Part_3', 'Q10_Part_4',
       'Q10_Part_5', 'Q10_Part_6', 'Q10_Part_7', 'Q10_Part_8', 'Q10_Part_9',
       'Q10_Part_10', 'Q10_Part_11', 'Q10_Part_12', 'Q10_Part_13',
       'Q10_Part_14', 'Q10_Part_15', 'Q10_Part_16', 'Q10_OTHER'],
      dtype='object')

-------------------------------------------

In [38]:
#1
columnas_data_txt_clean = np.array(archivo_txt.columns)
columnas_data_txt_clean

array(['index', 'Q3', 'Q4', 'Q5', 'Q6', 'Q8', 'Q11', 'Q13', 'Q15', 'Q20',
       'Q21', 'Q22', 'Q23', 'Q25', 'Q26', 'Q33', 'Q35', 'Q41', 'Q7', 'Q9',
       'Q12', 'Q14', 'Q16', 'Q17', 'Q24', 'Q32', 'Q34'], dtype=object)

In [39]:
#2
columnas_archivo_xml_clean = np.array(archivo_xml.columns)
columnas_archivo_xml_clean

array(['level_0', 'index', 'time', 'age', 'gender'], dtype=object)

In [40]:
#3
columnas_data_remaning_kaggle = np.array(data_remaning_kaggle.columns)
columnas_data_remaning_kaggle

array(['index', 'Q18_Part_1', 'Q18_Part_2', 'Q18_Part_3', 'Q18_Part_4',
       'Q18_Part_5', 'Q18_Part_6', 'Q18_OTHER', 'Q19_Part_1',
       'Q19_Part_2', 'Q19_Part_3', 'Q19_Part_4', 'Q19_Part_5',
       'Q19_OTHER', 'Q27_A_Part_1', 'Q27_A_Part_2', 'Q27_A_Part_3',
       'Q27_A_Part_4', 'Q27_A_Part_5', 'Q27_A_Part_6', 'Q27_A_Part_7',
       'Q27_A_Part_8', 'Q27_A_Part_9', 'Q27_A_Part_10', 'Q27_A_Part_11',
       'Q27_A_OTHER', 'Q28', 'Q29_A_Part_1', 'Q29_A_Part_2',
       'Q29_A_Part_3', 'Q29_A_Part_4', 'Q29_A_OTHER', 'Q30_A_Part_1',
       'Q30_A_Part_2', 'Q30_A_Part_3', 'Q30_A_Part_4', 'Q30_A_Part_5',
       'Q30_A_Part_6', 'Q30_A_Part_7', 'Q30_A_OTHER', 'Q31_A_Part_1',
       'Q31_A_Part_2', 'Q31_A_Part_3', 'Q31_A_Part_4', 'Q31_A_Part_5',
       'Q31_A_Part_6', 'Q31_A_Part_7', 'Q31_A_Part_8', 'Q31_A_Part_9',
       'Q31_A_OTHER', 'Q36_A_Part_1', 'Q36_A_Part_2', 'Q36_A_Part_3',
       'Q36_A_Part_4', 'Q36_A_Part_5', 'Q36_A_Part_6', 'Q36_A_Part_7',
       'Q36_A_OTHER', 'Q37_A_Part_1'

In [41]:
#4
columnas_exported_db_data = np.array(exported_db_data.columns)
columnas_exported_db_data

array(['index', 'Q10_Part_1', 'Q10_Part_2', 'Q10_Part_3', 'Q10_Part_4',
       'Q10_Part_5', 'Q10_Part_6', 'Q10_Part_7', 'Q10_Part_8',
       'Q10_Part_9', 'Q10_Part_10', 'Q10_Part_11', 'Q10_Part_12',
       'Q10_Part_13', 'Q10_Part_14', 'Q10_Part_15', 'Q10_Part_16',
       'Q10_OTHER'], dtype=object)

In [42]:
#combinacion 1,2
np.intersect1d(columnas_data_txt_clean, columnas_archivo_xml_clean)

array(['index'], dtype=object)

In [43]:
#combinacion 1,3
np.intersect1d(columnas_data_txt_clean, columnas_data_remaning_kaggle)

array(['index'], dtype=object)

In [44]:
#combinacion 1,4
np.intersect1d(columnas_data_txt_clean, columnas_exported_db_data)

array(['index'], dtype=object)

In [45]:
#combinacion 2,3
np.intersect1d(columnas_archivo_xml_clean, columnas_data_remaning_kaggle)

array(['index'], dtype=object)

In [46]:
#combinacion 2,4
np.intersect1d(columnas_archivo_xml_clean, columnas_exported_db_data)

array(['index'], dtype=object)

In [47]:
np.intersect1d(columnas_data_remaning_kaggle, columnas_exported_db_data)

array(['index'], dtype=object)

In [25]:
# 1 archivo_xml
# 2 archivo_txt
# 3 data_remaning_kaggle
# 4 exported_db_data

In [48]:
comb_1_2 = archivo_xml.merge(archivo_txt, how = 'inner', on = ['index'])
#comb_1_2.head()
#comb_1_2.columns

In [49]:
comb_1_2_3 = comb_1_2.merge(data_remaning_kaggle,how = 'inner', on = ['index'])
#comb_1_2_3.head()
#comb_1_2_3.columns

In [50]:
comb_1_2_3_4 = comb_1_2_3.merge(exported_db_data,how = 'inner', on = ['index'])
#comb_1_2_3_4.head()
#comb_1_2_3_4.columns

In [51]:
archivo_total = comb_1_2_3_4.copy()

In [53]:
archivo_total.to_csv("data/archivo_total.csv")

---

* Trabajamos en el fichero

In [54]:
archivo_trabajo_copia = archivo_total.copy()

In [55]:
archivo_trabajo_copia.shape

(25973, 260)

In [56]:
archivo_trabajo_copia.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25973 entries, 0 to 25972
Columns: 260 entries, level_0 to Q10_OTHER
dtypes: float64(8), int64(3), object(249)
memory usage: 51.7+ MB


/tmp/ipykernel_361/1550630736.py:1: FutureWarning: null_counts is deprecated. Use show_counts instead
  archivo_trabajo_copia.info(null_counts=True)


Comprobamos cuantos nulos tenemos.

In [57]:
archivo_trabajo_copia.isnull().sum()

level_0            0
index              0
time               0
age                0
gender             0
               ...  
Q10_Part_13    25722
Q10_Part_14    25713
Q10_Part_15    25826
Q10_Part_16    18799
Q10_OTHER      25236
Length: 260, dtype: int64

Comprobamos los principales estadístico de las columnas numéricas.

In [35]:
archivo_trabajo_copia.describe().T

count          mean            std    min     25%      50%  \
Unnamed: 0    25973.0  12986.000000    7497.903607    0.0  6493.0  12986.0   
level_0       25973.0  12986.000000    7497.903607    0.0  6493.0  12986.0   
index         25973.0  12986.000000    7497.903607    0.0  6493.0  12986.0   
time          25973.0  11054.664921  101471.622125  120.0   443.0    656.0   
Q30_B_Part_1      0.0           NaN            NaN    NaN     NaN      NaN   
Q30_B_Part_2      0.0           NaN            NaN    NaN     NaN      NaN   
Q30_B_Part_3      0.0           NaN            NaN    NaN     NaN      NaN   
Q30_B_Part_4      0.0           NaN            NaN    NaN     NaN      NaN   
Q30_B_Part_5      0.0           NaN            NaN    NaN     NaN      NaN   
Q30_B_Part_6      0.0           NaN            NaN    NaN     NaN      NaN   
Q30_B_Part_7      0.0           NaN            NaN    NaN     NaN      NaN   
Q30_B_OTHER       0.0           NaN            NaN    NaN     NaN      NaN   

                  75%        max  
Unnamed: 0    19479.0    25972.0  
level_0       19479.0    25972.0  
index         19479.0    25972.0  
time           1038.0  2488653.0  
Q30_B_Part_1      NaN        NaN  
Q30_B_Part_2      NaN        NaN  
Q30_B_Part_3      NaN        NaN  
Q30_B_Part_4      NaN        NaN  
Q30_B_Part_5      NaN        NaN  
Q30_B_Part_6      NaN        NaN  
Q30_B_Part_7      NaN        NaN  
Q30_B_OTHER       NaN        NaN

Nombre de las columnas.

In [58]:
archivo_trabajo_copia.columns

Index(['level_0', 'index', 'time', 'age', 'gender', 'Q3', 'Q4', 'Q5', 'Q6',
       'Q8',
       ...
       'Q10_Part_8', 'Q10_Part_9', 'Q10_Part_10', 'Q10_Part_11', 'Q10_Part_12',
       'Q10_Part_13', 'Q10_Part_14', 'Q10_Part_15', 'Q10_Part_16',
       'Q10_OTHER'],
      dtype='object', length=260)

Seleccionamos sólo las columnas necesarias para responder a nuestras preguntas en nuevo dataframe.

In [71]:
tabla_pequeña = archivo_trabajo_copia[['index','age', 'gender','Q3','Q4','Q5',"Q6","Q7","Q8","Q11","Q14","Q15","Q20","Q21","Q22","Q25","Q26","Q40_Part_1","Q40_Part_2","Q40_Part_3","Q40_Part_4","Q40_Part_5","Q40_Part_6","Q40_Part_7","Q40_Part_8","Q40_Part_9","Q40_Part_10","Q40_Part_11","Q40_OTHER","Q42_Part_1","Q42_Part_2","Q42_Part_3","Q42_Part_4","Q42_Part_5","Q42_Part_6","Q42_Part_7","Q42_Part_8","Q42_Part_9","Q42_Part_10","Q42_Part_11","Q42_OTHER"]].copy()
tabla_pequeña.head()

index    age gender         Q3                 Q4                       Q5  \
0      0  50-54    Man      India  Bachelor’s degree                    Other   
1      1  50-54    Man  Indonesia    Master’s degree  Program/Project Manager   
2      2  22-24    Man   Pakistan    Master’s degree        Software Engineer   
3      3  45-49    Man     Mexico    Doctoral degree       Research Scientist   
4      4  45-49    Man      India    Doctoral degree                    Other   

           Q6                  Q7      Q8  \
0  5-10 years           Python, R  Python   
1   20+ years   SQL, C, C++, Java  Python   
2   1-3 years   Python, C++, Java  Python   
3   20+ years              Python  Python   
4   < 1 years   Python, C, MATLAB  Python   

                                                 Q11  \
0                                           A laptop   
1  A cloud computing platform (AWS, Azure, GCP, h...   
2                                           A laptop   
3  A cloud computing platform (AWS, Azure, GCP, h...   
4  A cloud computing platform (AWS, Azure, GCP, h...   

                                                 Q14  \
0   Matplotlib ,Seaborn ,Ggplot / ggplot2 ,Shiny ...   
1                                        Matplotlib    
2                                        Matplotlib    
3                                        Matplotlib    
4             Matplotlib ,Seaborn ,Ggplot / ggplot2    

                                     Q15                        Q20  \
0                             5-10 years  Manufacturing/Fabrication   
1                               < 1 year  Manufacturing/Fabrication   
2  I do not use machine learning methods        Academics/Education   
3                             5-10 years        Academics/Education   
4                            10-20 years        Academics/Education   

                    Q21  Q22            Q25           Q26 Q40_Part_1  \
0      50-249 employees  3-4  25,000-29,999     $100-$999   Coursera   
1  1000-9,999 employees  1-2  60,000-69,999     $0 ($USD)        NaN   
2  1000-9,999 employees    0         $0-999     $0 ($USD)        NaN   
3  1000-9,999 employees    0  30,000-39,999     $0 ($USD)   Coursera   
4      50-249 employees  5-9  30,000-39,999  $1000-$9,999        NaN   

  Q40_Part_2            Q40_Part_3 Q40_Part_4 Q40_Part_5 Q40_Part_6  \
0        edX  Kaggle Learn Courses   DataCamp        NaN    Udacity   
1        NaN  Kaggle Learn Courses        NaN        NaN        NaN   
2        NaN                   NaN   DataCamp        NaN        NaN   
3        edX                   NaN        NaN        NaN        NaN   
4        NaN  Kaggle Learn Courses        NaN        NaN        NaN   

  Q40_Part_7 Q40_Part_8                                         Q40_Part_9  \
0      Udemy        NaN                                                NaN   
1        NaN        NaN  Cloud-certification programs (direct from AWS,...   
2        NaN        NaN                                                NaN   
3        NaN        NaN                                                NaN   
4      Udemy        NaN                                                NaN   

                                         Q40_Part_10 Q40_Part_11 Q40_OTHER  \
0                                                NaN         NaN       NaN   
1  University Courses (resulting in a university ...         NaN       NaN   
2                                                NaN         NaN       NaN   
3                                                NaN         NaN       NaN   
4                                                NaN         NaN       NaN   

  Q42_Part_1                                         Q42_Part_2 Q42_Part_3  \
0        NaN  Email newsletters (Data Elixir, O'Reilly Data ...        NaN   
1        NaN                                                NaN        NaN   
2        NaN                                                NaN        NaN   
3        NaN                      

In [60]:
tabla_pequeña.columns

Index(['index', 'age', 'gender', 'Q3', 'Q4', 'Q5', 'Q6', 'Q7', 'Q8', 'Q11',
       'Q14', 'Q15', 'Q20', 'Q21', 'Q22', 'Q25', 'Q26', 'Q40_Part_1',
       'Q40_Part_2', 'Q40_Part_3', 'Q40_Part_4', 'Q40_Part_5', 'Q40_Part_6',
       'Q40_Part_7', 'Q40_Part_8', 'Q40_Part_9', 'Q40_Part_10', 'Q40_Part_11',
       'Q40_OTHER', 'Q42_Part_1', 'Q42_Part_2', 'Q42_Part_3', 'Q42_Part_4',
       'Q42_Part_5', 'Q42_Part_6', 'Q42_Part_7', 'Q42_Part_8', 'Q42_Part_9',
       'Q42_Part_10', 'Q42_Part_11', 'Q42_OTHER'],
      dtype='object')

In [61]:
tabla_pequeña.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25973 entries, 0 to 25972
Data columns (total 41 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   index        25973 non-null  int64 
 1   age          25973 non-null  object
 2   gender       25973 non-null  object
 3   Q3           25973 non-null  object
 4   Q4           25973 non-null  object
 5   Q5           25973 non-null  object
 6   Q6           25973 non-null  object
 7   Q7           24941 non-null  object
 8   Q8           24940 non-null  object
 9   Q11          24720 non-null  object
 10  Q14          23970 non-null  object
 11  Q15          23754 non-null  object
 12  Q20          16325 non-null  object
 13  Q21          16251 non-null  object
 14  Q22          16126 non-null  object
 15  Q25          15391 non-null  object
 16  Q26          15183 non-null  object
 17  Q40_Part_1   9649 non-null   object
 18  Q40_Part_2   3028 non-null   object
 19  Q40_Part_3   8670 non-nul

Vamos a renombrar las columnas.

In [62]:
nombres_nuevos = {'index':"indices", 'age': "edad", 'gender':"genero", 'Q3':"pais", 'Q4': "formacion_2", 'Q5': "puesto_trabajo", 'Q6': "años_programando", 'Q7': "lenguajes_utilizados", 
                    'Q8':"lenguajes_recomendados", 'Q11': "herramientas_uso", 'Q14': "librerias_visualizacion", 'Q15': "años_usando_machine", 'Q20':"sector_trabajo_actual", 'Q21':"tamaño_empresa", 
                    'Q22':"personas_dpt_data", 'Q25':"salario_anual", 'Q26':"inversion_dpt_machine_5años", 'Q40_Part_1':"plataforma_coursera",'Q40_Part_2': "plataforma_edx", 
                    'Q40_Part_3': "plataforma_kaggle_learn_courses", 'Q40_Part_4': "plataforma_data_camp", 'Q40_Part_5': "plataforma_fast_ai", 'Q40_Part_6': "plataforma_udacity",
                    'Q40_Part_7': "plataforma_udemy", 'Q40_Part_8': "plataforma_linkedin_learning", 'Q40_Part_9':"plataforma_cloud_certification_programs", 'Q40_Part_10':"plataforma_university_courses",
                    'Q40_Part_11': "plataforma_none", 'Q40_OTHER': "plataforma_otros", 'Q42_Part_1':"fuente_conocimiento_twitter", 'Q42_Part_2':"fuente_conocimiento_email", 
                    'Q42_Part_3': "fuente_conocimiento_reddit", 'Q42_Part_4':"fuente_conocimiento_kaggle",'Q42_Part_5': "fuente_conocimiento_course",'Q42_Part_6': "fuente_conocimiento_youtube", 
                    'Q42_Part_7': "fuente_conocimiento_podcast", 'Q42_Part_8':"fuente_conocimiento_blogs", 'Q42_Part_9':"fuente_conocimiento_journal_publicacions",
                    'Q42_Part_10':"fuente_conocimiento_slack_communities", 'Q42_Part_11':"fuente_conocimiento_none", 'Q42_OTHER':"fuente_conocimiento_otros"}

In [63]:
tabla_pequeña.rename(columns = nombres_nuevos, inplace = True)

In [64]:
tabla_pequeña.columns

Index(['indices', 'edad', 'genero', 'pais', 'formacion_2', 'puesto_trabajo',
       'años_programando', 'lenguajes_utilizados', 'lenguajes_recomendados',
       'herramientas_uso', 'librerias_visualizacion', 'años_usando_machine',
       'sector_trabajo_actual', 'tamaño_empresa', 'personas_dpt_data',
       'salario_anual', 'inversion_dpt_machine_5años', 'plataforma_coursera',
       'plataforma_edx', 'plataforma_kaggle_learn_courses',
       'plataforma_data_camp', 'plataforma_fast_ai', 'plataforma_udacity',
       'plataforma_udemy', 'plataforma_linkedin_learning',
       'plataforma_cloud_certification_programs',
       'plataforma_university_courses', 'plataforma_none', 'plataforma_otros',
       'fuente_conocimiento_twitter', 'fuente_conocimiento_email',
       'fuente_conocimiento_reddit', 'fuente_conocimiento_kaggle',
       'fuente_conocimiento_course', 'fuente_conocimiento_youtube',
       'fuente_conocimiento_podcast', 'fuente_conocimiento_blogs',
       'fuente_conocimiento_

__________________________________________________________________________________________________________________________________________________________________________________________________

In [ ]:
"""
Salario medio en función de: Experiencia/ genero/ pais/ sector (q25,q2, q1, q3,q6)
Herramientas (plataformas, medios..) más utilizadas según: Experiencia/genero/sector/pais/edad (q1,q2,q3,q7,q11,q14,q20)
Lenguajes de programación en función de: genero/edad/pais/sector (q1,q2,q3,q7,q20)
Qué sectores son los que más demandan analítica de datos. (q1,q2,q3,q20,q21,q22)
Relación entre años de experiencia y tamaño de la empresa en la que trabaja, y nº de gente en el departamenteo, en base al genero, país y edad(q1,q2,q3,q6,q20,q21,q22)
Media de inversión en formación continua(q1,q2,q3,q5,q26,q40)

Que plataformas usas para estar al tanto de las nuevas del sector(q1,q2,q3,q42)"""

In [65]:
tabla_pequeña.isnull().sum()

indices                                         0
edad                                            0
genero                                          0
pais                                            0
formacion_2                                     0
puesto_trabajo                                  0
años_programando                                0
lenguajes_utilizados                         1032
lenguajes_recomendados                       1033
herramientas_uso                             1253
librerias_visualizacion                      2003
años_usando_machine                          2219
sector_trabajo_actual                        9648
tamaño_empresa                               9722
personas_dpt_data                            9847
salario_anual                               10582
inversion_dpt_machine_5años                 10790
plataforma_coursera                         16324
plataforma_edx                              22945
plataforma_kaggle_learn_courses             17303


In [66]:
tabla_pequeña.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25973 entries, 0 to 25972
Data columns (total 41 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   indices                                   25973 non-null  int64 
 1   edad                                      25973 non-null  object
 2   genero                                    25973 non-null  object
 3   pais                                      25973 non-null  object
 4   formacion_2                               25973 non-null  object
 5   puesto_trabajo                            25973 non-null  object
 6   años_programando                          25973 non-null  object
 7   lenguajes_utilizados                      24941 non-null  object
 8   lenguajes_recomendados                    24940 non-null  object
 9   herramientas_uso                          24720 non-null  object
 10  librerias_visualizacion                   2397

In [ ]:
#Las columnas de edad y sueldo deberían ser int

In [51]:
x= list(tabla_pequeña["edad"].unique())
x

['50-54',
 '22-24',
 '45-49',
 '25-29',
 '18-21',
 '30-34',
 '40-44',
 '35-39',
 '70+',
 '55-59',
 '60-69']

Decidimos crear una nueva columna con los países.

In [52]:
"""ASIA:
'India',
'Indonesia',
'Pakistan',
'Japan',
'China',
'Malaysia',
'Thailand',
'Nepal',
'Russia',
'Singapore',
'Turkey',
'Hong Kong (S.A.R.)',
'Iran, Islamic Republic of...',
'Viet Nam',
'Israel',
'Bangladesh',
'Tunisia',
 'Saudi Arabia',
'Taiwan',
'South Korea',
 'Philippines',
 'Sri Lanka',
 'United Arab Emirates',
'Kazakhstan',
 'Iraq',
AMERICA
'United States of America',
'Mexico',
'Brazil',
'Peru',
'Ecuador',
'Argentina',
'Colombia',
'Canada',
'Chile',
OCEANIA:
'Australia',
AFRICA
'Nigeria',
'South Africa',
'Ghana',
'Uganda',
'Egypt',
'Morocco',
'Algeria',
 'Kenya',
 'Ethiopia',
EUROPA:
'Greece',
'Belgium',
'Italy',
'Norway',
'Czech Republic'
'Poland',
'Spain',
'United Kingdom of Great Britain and Northern Ireland',
'France',
'Switzerland',
'Netherlands',
'Denmark',
'Germany',
'Sweden',
'Ireland',
'Portugal',
'Turkey',
'Ukraine',
 'Romania',
 'Austria',
'Belarus',
 'Other',
 'I do not wish to disclose my location',"""

"ASIA:\n'India',\n'Indonesia',\n'Pakistan',\n'Japan',\n'China',\n'Malaysia',\n'Thailand',\n'Nepal',\n'Russia',\n'Singapore',\n'Turkey',\n'Hong Kong (S.A.R.)',\n'Iran, Islamic Republic of...',\n'Viet Nam',\n'Israel',\n'Bangladesh',\n'Tunisia',\n 'Saudi Arabia',\n'Taiwan',\n'South Korea',\n 'Philippines',\n 'Sri Lanka',\n 'United Arab Emirates',\n'Kazakhstan',\n 'Iraq',\nAMERICA\n'United States of America',\n'Mexico',\n'Brazil',\n'Peru',\n'Ecuador',\n'Argentina',\n'Colombia',\n'Canada',\n'Chile',\nOCEANIA:\n'Australia',\nAFRICA\n'Nigeria',\n'South Africa',\n'Ghana',\n'Uganda',\n'Egypt',\n'Morocco',\n'Algeria',\n 'Kenya',\n 'Ethiopia',\nEUROPA:\n'Greece',\n'Belgium',\n'Italy',\n'Norway',\n'Czech Republic'\n'Poland',\n'Spain',\n'United Kingdom of Great Britain and Northern Ireland',\n'France',\n'Switzerland',\n'Netherlands',\n'Denmark',\n'Germany',\n'Sweden',\n'Ireland',\n'Portugal',\n'Turkey',\n'Ukraine',\n 'Romania',\n 'Austria',\n'Belarus',\n 'Other',\n 'I do not wish to disclose my loc